### PR TITLE
fix(manager/azure-pipelines): Enable azure repository extraction

### PR DESCRIPTION
**Pull request by tofay**
## Changes

Allow azure-pipelines manager to extract Azure DevOps (ADO) Git repositories from `resource.repository` entries in Azure pipeline definitions.

## Context
This is the first part of a fix for https://github.com/renovatebot/renovate/issues/15028.
It fixes the case where an ADO repository is referenced with its project name, but not the case where the repository entry implicitly references a repository in the same ADO project.

I'm happy to work on a fix for that if you can suggest an implementation suggestion - presumably the manager interface would need to change so manager's can determine details about the repository currently being updated

Partial fix for
- https://github.com/renovatebot/renovate/issues/15028

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 1]
```diff
@@ -13,7 +14,32 @@ const AzurePipelinesTaskRegex = regEx(/^(?<name>[^@]+)@(?<version>.*)$/);
...
+    // The repository URL is only deducible if we are running on AzureDevOps (so can use the endpoint)
+    // and the name is of the form `Project/Repository`.
+    // The name could just be the repository name, in which case AzureDevOps defaults to the
+    // same project, which is not currently accessible here. It could be deduced later by exposing
+    // the repository URL to managers.
+    // https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines#types
+    const { platform, endpoint } = GlobalConfig.get();
+    if (
+      platform === 'azure' &&
+      endpoint !== null &&
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:32]:
```suggestion
```
i think endpoint can never null or undefined at this stage.

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 2]
```diff
@@ -13,7 +14,32 @@ const AzurePipelinesTaskRegex = regEx(/^(?<name>[^@]+)@(?<version>.*)$/);
...
+      repository.name.includes('/')
+    ) {
+      const [projectName, repoName] = repository.name.split('/');
+      repositoryUrl = `${endpoint!}/${encodeURIComponent(
```
Comment (viceice) [inline lib/modules/manager/azure-pipelines/extract.ts:36]: please use `joinUrlParts` helper function instead

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/azure-pipelines/extract.ts [hunk 3]
```diff
@@ -13,7 +14,32 @@ const AzurePipelinesTaskRegex = regEx(/^(?<name>[^@]+)@(?<version>.*)$/);
...
+    // and the name is of the form `Project/Repository`.
+    // The name could just be the repository name, in which case AzureDevOps defaults to the
+    // same project, which is not currently accessible here. It could be deduced later by exposing
+    // the repository URL to managers.
+    // https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema/resources-repositories-repository?view=azure-pipelines#types
+    const { platform, endpoint } = GlobalConfig.get();
+    if (
+      platform === 'azure' &&
+      endpoint !== null &&
+      repository.name.includes('/')
```
Comment (JamieMagee) [inline lib/modules/manager/azure-pipelines/extract.ts:33]: Not necessary, but maybe add a debug log here that by adding the project name, Renovate would be able to address this.

**Review by JamieMagee — DISMISSED** [review]
Approved pending comments

File: lib/modules/manager/azure-pipelines/extract.ts (hunk 1 above)
Comment (tofay) [reply lib/modules/manager/azure-pipelines/extract.ts]: Looks like it can be undefined, but not null, so I've updated the code. thanks

**Review by viceice — APPROVED** [review]

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**